fix: allow doclist to have nested optional document (#1472)
docarray/docarray: docs/user_guide/representing/array.md

    To summarize: you should use `DocVec` when you need to work with contiguous data, and you should use `DocList` when you need to rearrange
    or extend your data.

## Dealing with Optional Field

Both [`DocList`][docarray.array.doc_list.doc_list.DocList] and [`DocVec`][docarray.array.doc_vec.doc_vec.DocVec] support optional fields but they behave differently.

!!! note
    This whole section is specific for nested BaseDoc, DocList and DocVec. 
    for other data type DocList and DocVec will treat the optional case as a normal case.

Let's take an example:

```python
from typing import Optional
from docarray.typing import NdArray
import numpy as np

class ImageDoc(BaseDoc):
    tensor: NdArray

class ArticleDoc(BaseDoc):
    image: Optional[ImageDoc]
    title: str
```

In this example `ArticleDoc` has an optional field `image` which is an `ImageDoc`. This means that this field can either
be None or be a `ImageDoc` instance.

> Remember for both DocList and DocVec calling `docs.image` will return a list like object of all the images of the documents.

Reviewer: ```suggestion
Remember that for both DocList and DocVec calling `docs.image` will return a list-like object of all the images of the documents.
```

> For DocList it will just iterate over all the documents and collect the image attribute of each document in a sequence for DocVec it will return the column of the image attribute.

Reviewer: ```suggestion
For DocList this call will iterate over all the documents and collect the image attribute of each document in a sequence, and for DocVec it will return the already stacked column of the `.image` attribute.
```

> The question which kind of sequence to you pick when the field is optional, i.e, some of the datapoint could be None ?

Reviewer: What does it mean "which kind of sequence to pick"?

For DocList it will return a list of `Optional[ImageDoc]` instead of a `DocList[ImageDoc]`:

```python
from docarray import DocList

docs = DocList[ArticleDoc](
    [
        ArticleDoc(image=ImageDoc(tensor=np.ones((3, 224, 224))), title="Hello"),
        ArticleDoc(image=None, title="World"),
    ]
)

assert docs.image == [ImageDoc(tensor=np.ones((3, 224, 224))), None]
```

> but for DocVec it is a bit different. Indeed, DocVec store the data for each filed as contiguous column. 

Reviewer: ```suggestion
But for DocVec it is a bit different. Indeed, a DocVec stores the data for each filed as contiguous column. 
```

> but for DocVec it is a bit different. Indeed, DocVec store the data for each filed as contiguous column. 

Reviewer: What does the fact that it stores it as a column to do with it not being a DocList? That column could in theory be a DocList, right?

Author: well not if some document are none ?

> This means that DocVec can create a column in only two case: either all the data for a field is None or all the data is not None.

Reviewer: ```suggestion
This means that DocVec can create a column in only two cases: either all the data for a field is None or all the data is not None.
```
I think in other places you have a name for this. Homogeneous? something like that? Would be good to introduce here I think.
Or did we take homogeneous to mean same doc type for every doc? Honestly I don't remember

For the first one the whole column will just be None. In the second The column will be a `DocList[ImageDoc]`

```python

> from docarray import DocVec

Reviewer: maybe use tabs for these examples?

Author: I split it a bit differently now

docs = DocVec[ArticleDoc](
    [
        ArticleDoc(image=ImageDoc(tensor=np.zeros((3, 224, 224))), title="Hello")
        for _ in range(10)
    ]
)
assert (docs.image.tensor == np.zeros((3, 224, 224))).all()
``` 

> Or it can be None

Reviewer: ```suggestion
Or it can be None:
```

    To summarize: you should use `DocVec` when you need to work with contiguous data, and you should use `DocList` when you need to rearrange
    or extend your data.

> ## Dealing with Optional fields

Author: yes

Both [`DocList`][docarray.array.doc_list.doc_list.DocList] and [`DocVec`][docarray.array.doc_vec.doc_vec.DocVec] support optional fields but they behave differently.

!!! note
    example of optional field
    
    ```python
    from typing import Optional
    from docarray import BaseDoc

    class MyDoc(BaseDoc):
        nested_doc: Optional[BaseDoc]
    ```

!!! note
    Using optional field is slightly different in some case so watch out. But in a nutshell :
    
    For nested BaseDoc 
    * DocList will return a list of document if the field is optional and a DocList if the field is not optional

>       * DocVec will return a DocList if all documents are there, or None if all docs are None. No mix of docs and None allowed!

Author: shit you are right ofc
